### ISSUE TITLE
Next.js project Cannot determine "minifiedUrl" argument for bugsnag-sourcemaps

### ISSUE THREAD

**Issue by sabrinaluohk01** [bugsnag/webpack-bugsnag-plugins#3]
Hello :)
bugsnag requires the publicPath, but in nextjs generated project, the public path is always with chunkhash, but the file is under `.next/bundle`.

So when I upload the sourcemap under `./next/bundle`, it wont work with the caught error in minified production code.

is there any other way to fix this issue? thanks!
### website:
![image](https://user-images.githubusercontent.com/36024221/37456011-66595c6c-2879-11e8-8242-1a19e9916c5b.png)

### error in bugsnag
![image](https://user-images.githubusercontent.com/36024221/37456055-7ab1703c-2879-11e8-9365-dc8ca08ae374.png)

### uploaded source map
![image](https://user-images.githubusercontent.com/36024221/37456086-8ea501a8-2879-11e8-88f9-64b6b4d3ecfd.png)

**Comment by bengourley** [comment]
Hi @sabrinaluohk01! Sorry to see that you're having issues.

I went to put together a small next.js example at to play around with things, and I had to fix a few issues with this library (#4). I've published that as `v1.1.1`.

Unfortunately that won't solve the issues you're having, though. The problem stems from the fact that the next bundles are not served with the same paths that they exist with on the filesystem…

The next router does [this](https://github.com/zeit/next.js/blob/eff27bdcfb7208e9cdadbfbeada4467384ffd379/server/index.js#L243-L274) – mapping paths like `/_next/xxxxxxxxxxxxxxx/page/index.js` onto `/.next/bundles/pages/index.js`. In order to support this kind of dynamic serving we'd need to add a feature to this library. At the moment we just join the `publicPath` with the location of the bundle on disk…

https://github.com/bugsnag/webpack-bugsnag-plugins/blob/07ffe2a62b91dd96571d75ce16fae9475a4e3a27/source-map-uploader-plugin.js#L54-L58

… assuming that the result of a webpack build is served from the filesystem statically – as it usually is.

Until we get around to supporting your use case, may I suggest an alternative. Rather than uploading sourcemaps to Bugsnag, you should leave the `//# sourceMappingURL=index.js.map` annotations in your JS, and allow Bugsnag's servers to retrieve the source map when we process an error. If your application isn't publicly available, you'll need to [whitelist our IPs](https://docs.bugsnag.com/platforms/browsers/source-maps/#access-through-firewall).

N.B. you should also add [this config](https://github.com/zeit/next.js/issues/3908#issuecomment-369873298) if you haven't already, to ensure source maps are created for all of your JS.

**Comment by azizhk** [comment]
Hey @bengourley, thanks for the help.
What do you suggest to upload server sourcemaps?

```js
webpack(config, { dev, isServer }) {
  if (!dev) {
    config.devtool = 'source-map'

    const appVersion = gitRev.short()
    const BUGSNAG_API = isServer
      ? dotenv.parsed.BUGSNAG_API_SERVER
      : dotenv.parsed.BUGSNAG_API_CLIENT

    config.plugins.push(
      new BugsnagBuildReporterPlugin(
        {
          apiKey: BUGSNAG_API,
          appVersion: appVersion,
          releaseStage: dotenv.parsed.BUGSNAG_ENV
        }
      )
    )
    config.plugins.push(
      new BugsnagSourceMapUploaderPlugin({
        apiKey: BUGSNAG_API,
        appVersion: appVersion,
        releaseStage: dotenv.parsed.BUGSNAG_ENV
      })
    )
  }
}
```
I started with something like this but came accross this issue. So now if I can upload only server side sourcemaps, that would be enough. I'll figure something else and comment for client side.

**Comment by bengourley** [comment]
@azizhk we don't currently support source maps for server-side javascript – however we do plan to support it soon.

**Comment by azizhk** [comment]
Hey, so here's my script to upload client side sourcemaps after a build. Works with next.js v6. Does not work with next.js v7. Thought would share.

```js
const fs = require('fs')
const _ = require('lodash')
const { resolve } = require('path')
const { promisify } = require('util')
const { upload: _uploadToBugsnag } = require('bugsnag-sourcemaps')
const { config: loadEnv } = require('dotenv')
const { short: gitRevShort } = require('git-rev-sync')

loadEnv({
  path: resolve(__dirname, '../.env')
})

const readFile = promisify(fs.readFile)
const uploadToBugsnag = promisify(_uploadToBugsnag)
const DASHBOARD_URL = process.env.DASHBOARD_URL

module.exports = async function uploadClientSourcemapTask(): Promise<void> {
  if (!process.env.BUGSNAG_API_CLIENT) {
    return
  }
  const BUILD_ID = await readFile(
    resolve(__dirname, '../.next/BUILD_ID'),
    'utf8'
  )
  const manifest = JSON.parse(
    await readFile(resolve(__dirname, '../.next/build-manifest.json'), 'utf8')
  )

  const baseConfig = {
    apiKey: process.env.BUGSNAG_API_CLIENT,
    appVersion: gitRevShort(),
    overwrite: true,
    uploadSources: true
  }

  const pageBundles = Object.keys(manifest)
    .filter(p => p.startsWith('bundles/pages'))
    .map(p => p.replace(/^bundles\/pages\//, ''))
  await Promise.all(
    pageBundles.map(p =>
      uploadToBugsnag({
        ...baseConfig,
        minifiedUrl: `${DASHBOARD_URL}_next/${BUILD_ID}/page/${p}`,
        minifiedFile: resolve(__dirname, `../.next/bundles/pages/${p}`),
        sourceMap: resolve(__dirname, `../.next/bundles/pages/${p}.map`)
      })
    )
  )

  const chunkBundles = _.entries(manifest).filter(([k, _]) =>
    k.startsWith('chunks')
  )

  await Promise.all(
    chunkBundles.map(([_, v]) =>
      uploadToBugsnag({
        ...baseConfig,
        minifiedUrl: `${DASHBOARD_URL}_next/webpack/${v[0]}`,
        minifiedFile: resolve(__dirname, `../.next/${v[0]}`),
        sourceMap: resolve(__dirname, `../.next/${v[0]}.map`)
      })
    )
  )

  await uploadToBugsnag({
    ...baseConfig,
    minifiedUrl: `${DASHBOARD_URL}_next/${manifest['main.js']}`,
    minifiedFile: resolve(__dirname, `../.next/${manifest['main.js']}`),
    sourceMap: resolve(__dirname, `../.next/${manifest['main.js']}.map`)
  })
}
```
Upgrading to next 7 soon, will share if successful.

**Comment by azizhk** [comment]
With Next 7, this plugin works as is and does not need much modifications. As next now stores the files in the same path that their served to the client.

Just you need to turn on sourcemaps. For that you can use `@zeit/next-source-maps`, just need https://github.com/zeit/next-plugins/pull/296 to get merged.

Till then you can duplicate https://github.com/zeit/next-plugins/pull/296 's code in your config

next.config.js:
```js
const withSourceMaps = require('@zeit/next-source-maps')
module.exports = withSourceMaps({
  webpack(config, options) {
    const { dev, isServer, defaultLoaders } = options
    if (!dev && !isServer) {
      config.devtool = 'hidden-source-map'
      if (config.optimization && config.optimization.minimizer) {
        for (const plugin of config.optimization.minimizer) {
          if (plugin.constructor.name === 'TerserPlugin') {
            plugin.options.sourceMap = true
            break
          }
        }
      }

      config.plugins.push(
        new BugsnagSourceMapUploaderPlugin({
          apiKey: BUGSNAG_API_KEY, // replace this
          publicPath: `http*://*.example.com/_next/` // replace this
        })
      )
    }
  }
})
```

Above code is for `@zeit/next-source-maps` version 0.0.3
You should be able to close this issue now.

**Comment by bnisbett** [comment]
@bengourley Follow up question from earlier in the thread. Is there now support for source maps for server-side javascript?

**Comment by isBatak** [comment]
@bengourley could we reopen this issue?
I think we need a way to define a wildcard in place of next BUILD_ID.
For example `minifiedUrl` like this one:
`https://example.com/_next/static/L0eJ6XV0Hw9IEdAUlJFh5/pages/example.js`
should be:
`https://example.com/_next/static/*/pages/example.js`
I'm thinking to add an additional `transformSource` option to plugin setting that will allow transforming of a source string, e.g. `static/L0eJ6XV0Hw9IEdAUlJFh5/pages/example.js` to `static/*/pages/example.js`.

**Comment by snmaynard** [comment]
@bnisbett yes we support server side source maps now! You should upload your source maps -> https://docs.bugsnag.com/platforms/javascript/source-maps/#source-map-upload

**Comment by azizhk** [comment]
So here is my next.config.js to upload client and server sourcemaps.
Next.js version `8`
@zeit/next-source-maps version `0.0.4-canary.1`

```js
const fp = require("lodash/fp");

const enhance = fp.compose(
  require("@zeit/next-source-maps")({ devtool: "hidden-source-map" }),
  // Add other plugins here
);

module.exports = enhance({
  webpack(config, { isServer, dev }) {
    if (isServer) {
      config.devtool = "source-map";
    }

    const BUGSNAG_API = isServer
      ? BUGSNAG_API_SERVER_KEY // replace here
      : BUGSNAG_API_CLIENT_KEY; // replace here
    if (!dev) {
      const { BugsnagSourceMapUploaderPlugin } = require("webpack-bugsnag-plugins");

      const SELF_URL = "https://example.com/"; // replace here
      config.plugins.push(
        new BugsnagSourceMapUploaderPlugin({
          apiKey: BUGSNAG_API,
          overwrite: true,
          releaseStage: "production",
          publicPath: isServer
            ? ".next/server/"
            : `${SELF_URL}_next/`
        })
      );
    }

    return config;
  }
});
```

I haven't encountered an error yet so can't verify whether this works. But I am uploading the sourcemap to the correct path as it comes up in the stacktrace. (See highlighted selected part, sorry for the blurring)
<img width="1146" alt="Screen Shot 2019-04-24 at 9" src="https://user-images.githubusercontent.com/5094296/56630909-c17bd800-6670-11e9-8989-ce2b4c6049c8.png">

**Comment by lifeiscontent** [comment]
@bengourley do you think you could share how to integrate this into nextjs? the example doesn't talk about how to get sourcemaps working.

**Comment by azizhk** [comment]
My solution above works for me. @lifeiscontent can you mention your next.js version and I can point you accordingly.

**Comment by dcalhoun** [comment]
@lifeiscontent the following is my config for Next 7 that successfully loads source maps to Bugsnag. It is based off the work @azizhk already shared.

```js
const {
  bugsnagApiKeyServer,
  bugsnagApiKeyClient,
} = require("./env-config.js");
const {
  BugsnagBuildReporterPlugin,
  BugsnagSourceMapUploaderPlugin
} = require("webpack-bugsnag-plugins");

const DEPLOY_ENV = process.env.DEPLOY_ENV || "development";

module.exports = {
  serverRuntimeConfig: { bugsnagApiKey: bugsnagApiKeyServer },
  publicRuntimeConfig: { bugsnagApiKey: bugsnagApiKeyClient },
  webpack: (config, { buildId, isServer }) => {
    if (DEPLOY_ENV !== "development") {
      config.plugins.push(
        new BugsnagBuildReporterPlugin({
          apiKey: isServer ? bugsnagApiKeyServer : bugsnagApiKeyClient,
          appVersion: buildId,
          releaseStage: DEPLOY_ENV,
          sourceControl: {
            provider: "<PROVIDER_NAME>",
            repository: "<PROVIDER_URL>",
            revision: "$Format:%H$"
          }
        })
      );

      config.plugins.push(
        new BugsnagSourceMapUploaderPlugin({
          apiKey: isServer ? bugsnagApiKeyServer : bugsnagApiKeyClient,
          appVersion: buildId,
          overwrite: true,
          publicPath: isServer ? ".next/server" : `<DOMAIN>/_next/`
        })
      );
    }

    return config;
  }
};
```

**Comment by lifeiscontent** [comment]
@azizhk here's my config that is currently not working

``` js
const fs = require('fs');
const path = require('path');
const withSass = require('@zeit/next-sass');
const withSourceMaps = require('@zeit/next-source-maps');
const {
  BugsnagBuildReporterPlugin,
  BugsnagSourceMapUploaderPlugin,
} = require('webpack-bugsnag-plugins');

const RELEASE_STAGE = process.env.RELEASE_STAGE || 'development';

function getRevision() {
  if (fs.existsSync('.git')) {
    const rev = fs
      .readFileSync('.git/HEAD')
      .toString()
      .trim()
      .split(/.*[: ]/)
      .slice(-1)[0];
    if (rev.indexOf('/') === -1) {
      return rev;
    } else {
      return fs
        .readFileSync(`.git/${rev}`)
        .toString()
        .trim();
    }
  }
}

module.exports = withSourceMaps(
  withSass({
    publicRuntimeConfig: {
      BUGSNAG_API_KEY: process.env.BUGSNAG_BROWSER_API_KEY,
    },
    serverRuntimeConfig: {
      BUGSNAG_API_KEY: process.env.BUGSNAG_SERVER_API_KEY,
    },
    sassLoaderOptions: {
      includePaths: ['styles', 'node_modules'].map(d =>
        path.join(__dirname, d)
      ),
      sourceMap: true,
    },
    cssLoaderOptions: {
      sourceMap: true,
    },
    webpack(config, { buildId, isServer }) {
      if (RELEASE_STAGE !== 'development') {
        config.plugins.push(
          new BugsnagBuildReporterPlugin({
            apiKey: isServer
              ? process.env.BUGSNAG_SERVER_API_KEY
              : process.env.BUGSNAG_BROWSER_API_KEY,
            appVersion: process.env.HEROKU_RELEASE_VERSION || buildId,
            releaseStage: RELEASE_STAGE,
            notifyReleaseStages: ['staging', 'production'],
            sourceControl: {
              provider: 'github',
              repository: 'https://github.com/redacted/redacted',
              revision: getRevision(),
            },
          })
        );

        config.plugins.push(
          new BugsnagSourceMapUploaderPlugin({
            apiKey: isServer
              ? process.env.BUGSNAG_SERVER_API_KEY
              : process.env.BUGSNAG_BROWSER_API_KEY,
            appVersion: process.env.HEROKU_RELEASE_VERSION || buildId,
            overwrite: true,
            publicPath: isServer ? '.next/server' : '*/_next/',
            releaseStage: RELEASE_STAGE,
            notifyReleaseStages: ['staging', 'production'],
          })
        );
      }
      return config;
    },
  })
);
```

**Comment by azizhk** [comment]
Can you share the version of next.js that you are using?

**Comment by mattdyoung** [comment]
Hi @lifeiscontent 

Thanks for raising the related support request. As mentioned on the response, looking at your account I can see that your uploaded source maps have an app version specified but the error reports we're receiving from your application don't, so setting the `appVersion` when initializing Bugsnag in your application should resolve this.

**Comment by lifeiscontent** [comment]
@azizhk I'm using next 8 and it seems source maps are not being uploaded on the browser end of things

**Comment by mattdyoung** [comment]
This ended up being resolved under a Bugsnag support request.

For anyone else with this issue, it looks to be an issue with Next, but we've found a workaround.

You would need to apply this package:
https://github.com/zeit/next-plugins/tree/master/packages/next-source-maps
but you'll need the latest pre-release version:
`@zeit/next-source-maps@0.0.4-canary.1`

Next doesn't output source maps by default. This plugin enables source map generation, but with versions before this one it had a bug where it didn't output for browser, only for Node.js. Here's the related issue:
https://github.com/zeit/next-plugins/issues/377#issuecomment-513604330

**Comment by VadymBoguslavsky** [comment]
You mean here?  

```
import React from 'react';
import bugsnag from '@bugsnag/js';
import bugsnagReact from '@bugsnag/plugin-react';

const bugsnagClient = bugsnag({
  apiKey: 'd0a59e8349ee3e1ba609cdbd2ecaecc5',
  notifyReleaseStages: ['production', 'staging'],
  appVersion: '1.0.0'  <---- is this it?
});

bugsnagClient.use(bugsnagReact, React);

export default bugsnagClient;

```

as I understand it buildId and this version must be the same, but how to know buildID?

**Comment by azizhk** [comment]
```js
if (typeof window !== "undefined" && window.__NEXT_DATA__ && window.__NEXT_DATA__.buildId) {
  const bugsnagClient = bugsnag({
    apiKey: 'd0a59e8349ee3e1ba609cdbd2ecaecc5',
    notifyReleaseStages: ['production', 'staging'],
    appVersion: window.__NEXT_DATA__.buildId
  });
}
```

**Comment by azizhk** [comment]
> when i starting it locally it buildId: "development"
Yeah when you build in production mode, it will generate and populate a buildId.

Regarding the buildId on server, you can use webpack's define plugin:
```js
// next.config.js
module.exports = {
  webpack: (config, { buildId, dev, isServer, defaultLoaders }) => {
    config.plugins.push(
      new webpack.DefinePlugin({
        "process.env.APP_VERSION": JSON.stringify(buildId),
      })
    );
  }
}
```

This will give you the buildId in `process.env.APP_VERSION`

**Comment by VadymBoguslavsky** [comment]
Sorry but it seems that it`s all the same(

![Screen Shot 2019-09-13 at 2 12 23 PM](https://user-images.githubusercontent.com/22998957/64858680-bc73f100-d630-11e9-810d-a728d9d5b3be.png)

Now in config:

```
 config.plugins.push(
          new webpack.DefinePlugin({
            "process.env.APP_VERSION": JSON.stringify(buildId),
          })
        );

        if (!dev) {
          config.plugins.push(
            new BugsnagSourceMapUploaderPlugin({
              apiKey: '**************',
              appVersion: process.env.APP_VERSION,
              publicPath: '*/',
              override: true,
              uploadSource: true
            })
          );
        } 
```


and in bugsnag config is

```
import React from 'react';
import bugsnag from '@bugsnag/js';
import bugsnagReact from '@bugsnag/plugin-react';
import config from '../config';

const bugsnagClient = bugsnag({
  apiKey: config.apiKey,
  notifyReleaseStages: ['production', 'staging'],
  appVersion: process.browser
    ? global.__NEXT_DATA__.buildId
    : process.env.APP_VERSION
});

bugsnagClient.use(bugsnagReact, React);

export default bugsnagClient;
```

**Comment by azizhk** [comment]
No need to do 
```
  appVersion: process.browser
    ? global.__NEXT_DATA__.buildId
    : process.env.APP_VERSION
```
Now you can directly use `process.env.APP_VERSION` for both client and server.

**Comment by mattdyoung** [comment]
@VadymBoguslavsky 
I'd suggest if you're still having issues you contact support@bugsnag.com with full details of your config and a link to an example of an error in your dashboard which hasn't correctly mapped which will enable us to analyze the root cause much more quickly.

**Comment by lifeiscontent** [comment]
@mattdyoung seems a recent version of next.js broke sourcemaps again. Should I reach out to support@bugsnag.com?

**Comment by phillipsam** [comment]
Hey @lifeiscontent if you're ever experiencing any issues, please don't hesitate to reach out to support@bugsnag.com!

**Comment by zhirzh** [comment]
same problem exists in nextjs 14 with server pages sourcemaps. looking inside the [compilation `stats` object](https://github.com/bugsnag/webpack-bugsnag-plugins/blob/v1.8.0/source-map-uploader-plugin.js#L41), i found that page chunk files and auxiliary files are resolved relative to `outputPath` which is set to `.next/server/chunks`:
```jsonc
{
  "publicPath": "/_next/",
  "outputPath": "/path/to/project/.next/server/chunks",
  "chunks": [
    {
      "names": ["pages/hello"],
      "files": ["../pages/hello.js"],
      "auxiliaryFiles": ["../pages/hello.js.map"],
      // ...
    },
    {
      "reason": "split chunk (cache group: default)",
      "names": [],
      "files": ["123.js"],
      "auxiliaryFiles": ["123.js.map"],
      // ...
    },
    // ...
  ],
  // ...
}
```

this causes sourcemaps to be uploaded with `".."` in the url and bugsnag cannot match them to the actual pages that report issues:

```
<i> [BugsnagSourceMapUploaderPlugin] uploading sourcemap for ".next/server/chunks/../pages/hello.js"
<i> [BugsnagSourceMapUploaderPlugin] uploading sourcemap for ".next/server/chunks/123.js"
```

---

i was able to "fix" this by deleting parent directory reference (`/chunks/../` -> `/`). this simple change worked for my problem but i couldn't think of a more general solution.

PS: i also needed to pass `publicPath` to the plugin config (as @azizhk pointed out [here](https://github.com/bugsnag/webpack-bugsnag-plugins/issues/3#issuecomment-530818582)) and use different value for server builds

```js
new BugsnagSourceMapUploaderPlugin({
  // ...
  publicPath: isServer ? ".next/server/chunks/" : "*/_next/"
})
```


<details>
<summary>
here is the patch for <a href="https://github.com/ds300/patch-package">patch-package</a>:
</summary>

```diff
diff --git a/node_modules/webpack-bugsnag-plugins/source-map-uploader-plugin.js b/node_modules/webpack-bugsnag-plugins/source-map-uploader-plugin.js
index 8d77345..0d27417 100644
--- a/node_modules/webpack-bugsnag-plugins/source-map-uploader-plugin.js
+++ b/node_modules/webpack-bugsnag-plugins/source-map-uploader-plugin.js
@@ -79,14 +79,19 @@ class BugsnagSourceMapUploaderPlugin {
             return null
           }
 
+          let url = '' +
+            // ensure publicPath has a trailing slash
+            publicPath.replace(/[^/]$/, '$&/') +
+            // remove leading / or ./ from source
+            source.replace(/^\.?\//, '')
+
+          // replace parent directory references with empty string
+          url = url.replace(/[^\/]+\/\.\.\//g, '')
+
           return {
             source: outputChunkLocation,
             map: outputSourceMapLocation,
-            url: '' +
-              // ensure publicPath has a trailing slash
-              publicPath.replace(/[^/]$/, '$&/') +
-              // remove leading / or ./ from source
-              source.replace(/^\.?\//, '')
+            url
           }
         }).filter(Boolean)
       }
```

</details>

**Comment by hannah-smartbear** [comment]
Hi @zhirzh,

I just wanted to see if you are still experiencing any issues uploading source maps with the correct URL or if your workaround has fixed this for you? If you are still experiencing issues, are you able to send us a screenshot of the URL shown on your BugSnag events that you are trying to match when uploading source maps? Or, even better, are you able to write in to us at [support@bugsnag.com](mailto:support@bugsnag.com) with a link to an example event in your dashboard?

**Comment by micahsteinberg** [comment]
Hi @hannah-smartbear ,

I'm facing this issue too. Here are screenshots of the URLs shown in our bugnsag events with `..`s, and of an example event with the source map not working.

![Image](https://github.com/user-attachments/assets/9f3331a3-0a91-409c-8ef6-01c13bf5cab5)

![Image](https://github.com/user-attachments/assets/635ae66f-5c2a-466b-9bc0-dee35a65df07)

I tested using @zhirzh patch locally and it fixes the issue, but would much rather an actual long term fix > using patch package. Let me know if you have any ideas and can help?

**Comment by mclack** [comment]
Hi @micahsteinberg 

We have scheduled some improvements in the `webpack-bugsnag-plugins` which should alleviate this issue, and avoid the need for a patch package. We will make sure to update this thread once a new release including these improvements is available for you to try.

**Comment by mclack** [comment]
Hi all,

We have now released v1.9.0 of `webpack-bugsnag-plugins` which includes handling for parent directory references in source paths: https://github.com/bugsnag/webpack-bugsnag-plugins/releases/tag/v1.9.0

Please let us know if you have any issues or questions with this after upgrading. You can open a ticket with us directly by contacting support@bugsnag.com
